### PR TITLE
Add the possibility to disallow unknown channels to avoid DoS

### PR DESCRIPTION
**Pull request by maxatome**
Until an authorization framework comes up for limiting subscribing (à
la nats), add a store_limits parameter: unknown_channels_disallowed.
When set to true, only channels listed in store_limits.channels hash
are allowed to be subscribed.

### CODE REVIEW

**Comment by kozlovic** [comment]
Hello Maxime!
Thank you for this PR: this is something we were planning to do actually for a possible FT with partitioning of channels among groups of FT servers. 

That being said, I don't think the PR could be merged as-is: my guess, looking at the code, it that all messages - regardless of their channel - would be rejected if you enable the new parameter. The reason I say that is that I see that you added code in the server's lookup and if the option is enabled, you simply reject. I don't see where you would do the lookup to check if the channel is known or not.
I think that there are 2 ways in solving that: move the check in the store implementation, which right now would be in `stores/common.go` and is shared by memory and file stores, but would maybe need to be replicated if we were to add new stores, or do the check in the server itself, but it means that we should make a copy of the PerChannel map for the server's use.

As I said, we may need to do that shortly, so it is up to you to see if you still want to experiment.
Note that we have just recently published a contributions guidelines that you can find [here](http://nats.io/documentation/CONTRIBUTING/). Also, in very near future, you should have a pull request template opening when trying to create a PR. It would look like [this](https://github.com/nats-io/gnatsd/blob/master/.github/PULL_REQUEST_TEMPLATE.md). For instance, your PR lack some testing and code coverage. Adding a test that checks that some channels are authorized while others are not would be a minimum and would have show that the PR is not working as expected.

We appreciate you trying to contribute and see the value of adding such feature. Thanks again.

**Comment by maxatome** [comment]
Hello,
You are right. When I did my tests, all channels have been already created by the server, that's why it worked :(

I have a new implementation that seems to work using a third way:
- add a `IsPredeclaredChannel(channel string) bool` to the `stores.Store` interface
- implement it for `genericStore` testing whether `gs.limits.PerChannel[channel]` exists or not
- still testing in `StanServer.lookupOrCreateChannel()` but using `s.store.IsPredeclaredChannel()` in addition to `s.opts.UnknownChannelsDisallowed`

This way we can avoid copying `PerChannel` map, is'nt it? Perhaps I miss something.

I pushed these modifications in my branch, I have to complete the unit tests before pull-requesting again following your guidelines. Tell me whether you are interested or not.

Thanks for your work.

**Comment by kozlovic** [comment]
@maxatome 

I see value in what you are trying to do, and really think that we will need something similar for FT with partitioning. But then, I think I would need to have the info readily available in the server anyway. The problem with the approach you are taking is that it requires a change in the Store interface, which some users have already some store implementations of their own. That would be fine to change the interface (I am actually doing just that in #263), but if I soon need to have access to the list of channels in the server code itself, it would be not ideal to rip it off from the store interface soon after it was added. The other drawback of having the check in the store is, as I said, needing to possibly replicate the check in all store implementations: as you can see, right now it is in the common store implementation, but not all stores may inherit from the generic store.

So in short, I would say that for now you can proceed on your fork, but I would not consider merging until I am progressing a bit more in the channels partitioning idea that will follow the FT work for #263. Is that fair?

As a side note, your latests changes in the PR failed because of the tool we use to do some checks was temporarily broken in that they were using Go 1.8 specific code, which they reverted. I have updated travis and fixed some errors reported by those tools. When you do add to your branch, I would recommend that you first rebase with master (I have merged #267).

Thanks again for your contribution!

**Comment by maxatome** [comment]
@kozlovic I understand, that's why I asked you whether you are interested or not. No problem :)
I developed this feature because I need it to avoid service abuse.

Note that we can achieve the same result without modifying any code, by reducing the store_limits.max_channels to the number of store_limits.channels items and then immediately subscribe one time to each of them, with the difference that the error message received by the client subscribing to a wrong channel is a bit confusing (too many channels), but it's a detail.

Thanks for your time, I let you close this pull request.

**Comment by kozlovic** [comment]
@maxatome Thank you for your understanding. As I said, I may need to use the list of channels on startup to have internal subscriptions on these subjects only (for channels partitioning on server servers with same cluster ID). If that is the case, servers would only receive messages on those channels anyway, so there would not be a need to check. Since I would need/use the list in the server and not the store, even without the partitioning feature, it would then be easy for the server to check in `lookupOrCreateChannel`.

So I am closing this PR now but will be sure to reference it/mention it when implementing the other approach.

Thanks again for your contribution and using NATS Streaming!
